pkg/yurthub/otaupdate/ota.go

```diff
@@ -87,29 +97,48 @@ func UpdatePod(clientset kubernetes.Interface, nodeName string) http.Handler {
 		namespace := params["ns"]
 		podName := params["podname"]
 
-		err, ok := applyUpdate(clientset, namespace, podName, nodeName)
-		// Pod update failed with error
-		if err != nil {
-			WriteErr(w, "Apply update failed", http.StatusInternalServerError)
-			return
-		}
+		pod, ok := preCheck(clientset, namespace, podName, nodeName)
 		// Pod update is not allowed
 		if !ok {
-			WriteErr(w, "Pod is not-updatable", http.StatusForbidden)
+			util.WriteErr(w, "Pod is not-updatable", http.StatusForbidden)
+			return
+		}
+
+		var upgrader OTAUpgrader
+		kind := pod.GetOwnerReferences()[0].Kind
+		switch kind {
+		case StaticPod:
+			upgrader = &static.StaticPodUpgrader{Interface: clientset,
+				NamespacedName: types.NamespacedName{Namespace: namespace, Name: podName}}
+		case DaemonPod:
+			upgrader = &daemon.DaemonPodUpgrader{Interface: clientset,
+				NamespacedName: types.NamespacedName{Namespace: namespace, Name: podName}}
+		default:
+			util.WriteErr(w, fmt.Sprintf("Not support ota upgrade pod type %v", kind), http.StatusBadRequest)
+			return
+		}
+
+		if err := upgrader.Apply(); err != nil {
+			// Pod update failed with error
+			util.WriteErr(w, "Apply update failed", http.StatusInternalServerError)
 			return
 		}
 
 		// Successfully apply update, response 200
-		WriteJSONResponse(w, []byte(fmt.Sprintf("Start updating pod %v/%v", namespace, podName)))
+		util.WriteJSONResponse(w, []byte(fmt.Sprintf("Start updating pod %v/%v", namespace, podName)))
 	})
 }
 
-// applyUpdate execute pod update process by deleting pod under OnDelete update strategy
-func applyUpdate(clientset kubernetes.Interface, namespace, podName, nodeName string) (error, bool) {
+// preCheck will check the necessary requirements before apply upgrade
+// 1. target pod has not been deleted yet
+// 2. target pod belongs to current node
+// 3. check whether target pod is updatable
+// At last, return the target pod to do further operation
+func preCheck(clientset kubernetes.Interface, namespace, podName, nodeName string) (*corev1.Pod, bool) {
```

Comment:
I think we can add static pod configmap loading as one of check items in preCheck.

Reply:
`preCheck` function is used to check whether the given pod can be upgraded or not. Configmap has nothing to do with this check. StaticPod controller will make sure the relevant configmap existed in the cluster. 

If you think necessary, we can use dynamic client to get targeted CR and then generate the manifest directly when we cannot find the configmap.

Reply:
@xavier-hou I don't agree with you, and i think that configmap loading is one part of static pod upgrading, because if we fail to download the configmap, the static pod upgrade should not proceed.

Reply:
I prefer to check the existence of configmap in `Switch-case` struct. 
1. Daemon pods do not need this check.
2. It's hard to judge the pod type before get this pod and `preCheck` is also used to get the pod in the previous implementation.